text/0000-async-project-goal-2024h2.md

```diff
@@ -0,0 +1,275 @@
+
+[AVD]: https://rust-lang.github.io/wg-async/vision.html
+
+This RFC will lay out a "plan of attack" for async, including both obvious good things (similar to [async closures][]) but also "known unknowns" and ways to resolve them. Areas the RFC is expected to cover are as follows:
+
+[Making Async Rust Reliable]: https://tmandry.gitlab.io/blog/posts/making-async-reliable/
+
+* Status quo, covering biggest challenges
+    * Lack of strong learning material
+    * Common idioms contain footguns that cause unexpected failures (see e.g., Tyler's blog post [Making Async Rust Reliable][])
+    * Low-level performance hurdles, such as large future sizes and downsides of the poll model
+    * Fragmentation between runtimes
+* Design axioms to pursue for async (see e.g. axioms proposed)
+* Goals, some variant of
+    * Free of accidental complexity
+    * Easy to get started
+    * Easy to pick executor and integrate with other systems (e.g., mobile runtimes, company-specific threadpools, etc)
+    * Moderately easy to adapt to "extreme" embedded environments
+    * Good performance by default, peak performance with tuning
+* Key unknowns in terms of how to achieve the above goals, for example 
+    * how to replace footgun-prone APIs with more reliable alternatives:
+        * buffered-streams, cancellation (esp. due to use of select)
+        * patterns to express
+            * merged streams -- processing one stream of data with occasional control events
+            * task parallelism
+        * cleanup and teardown
+            * ordered destruction
+    * how should async drop work (`?Leak` vs `?Drop` vs whatever):
+        * how to prevent async drop from occuring in sync contexts?
+    * what does runtime interface look like?
+        * Can/should we be generic over runtime
+* Strategy for how to get where we are going
+    * What problems to attack first
+    * How to reduce or find solutions to the above unknowns
+
+### Complete async drop experiments
+
+Authors of async code frequently need to call async functions as part of resource cleanup. Because Rust today only supports synchronous destructors, this cleanup must take place using alternative mechanisms, forcing a divergence between sync Rust (which uses destructors to arrange cleanup) and async Rust. [MCP 727][] proposed a series of experiments aimed at supporting async drop in the compiler. We would like to continue and complete those experiments. These experiments are aimed at defining how support for async drop will be implemented in the compiler and some possible ways that we could modify the type system to support it (in particular, one key question is how to prevent types that whose drop is async from being dropped in sync code).
```

Comment:
```suggestion
Authors of async code frequently need to call async functions as part of resource cleanup. Because Rust today only supports synchronous destructors, this cleanup must take place using alternative mechanisms, forcing a divergence between sync Rust (which uses destructors to arrange cleanup) and async Rust. [MCP 727][] proposed a series of experiments aimed at supporting async drop in the compiler. We would like to continue and complete those experiments. These experiments are aimed at defining how support for async drop will be implemented in the compiler and some possible ways that we could modify the type system to support it (in particular, one key question is how to prevent types whose drop is async from being dropped in sync code).
```